packages/babel-parser/src/parse-error/credentials.js

```diff
@@ -0,0 +1,66 @@
+// @flow
+
+export const ParseErrorCodes = Object.freeze({
+  SyntaxError: "BABEL_PARSER_SYNTAX_ERROR",
+  SourceTypeModuleError: "BABEL_PARSER_SOURCETYPE_MODULE_REQUIRED",
+});
+
+export type ParseErrorCode = $Values<typeof ParseErrorCodes>;
+
+export type SyntaxPlugin =
+  | "flow"
+  | "typescript"
+  | "jsx"
+  | "pipelineOperator"
+  | "placeholders";
+
+export type ToMessage<ErrorDetails> = (self: ErrorDetails) => string;
+
+export type ParseErrorCredentials<ErrorDetails> = {
+  code: ParseErrorCode,
+  reasonCode: string,
+  syntaxPlugin?: SyntaxPlugin,
+
+  toMessage: ToMessage<ErrorDetails>,
+};
+
+const thisify = <T>(f: (self: any, ...args: any[]) => T): ((...any[]) => T) =>
+  function (...args: any[]): T {
+    return ((f(this, ...args): any): T);
+  };
+
+const reflect = (keys: string[], last = keys.length - 1) => ({
+  get: self => keys.reduce((object, key) => object[key], self),
+  set: (self, value) =>
```

Comment:
I'd prefer to just define descriptors using a real `this`, and avoid both `thisify` and the `...(descriptor.get && { get: thisify<any>(descriptor.get) })`s below.

Reply:
Yup, sorry got stubborn working on this at night, will do.

Reply:
Done: https://github.com/babel/babel/pull/14320/commits/97ab2abdbc4974328fd49cecf6bd7085f9dfff11